sycl/test/basic_tests/fpga_device_link_diag.cpp

```diff
@@ -1,27 +0,0 @@
-//==----- fpga_device_link_diag.cpp - SYCL FPGA linking diagnostic test ----==//
```

Comment:
If I understand correctly, the general idea here is that the whole `llvm-no-spir-kernel` routine becomes irrelevant now that incremental device linking is allowed (which I personally consider great news, BTW). There are a couple additional points to consider:
1. Could users run into any caveats with backwards incompatibility of compiler/driver behaviors? Not that I could come up with any myself, but a confirmation from you as the implementer & expert would be greatly appreciated.
2. Provided that there aren't any pitfalls with **1**, we could consider removing `llvm-no-spir-kernel` as such. I believe this could be done separately, as all code that actually uses the tool is removed with your patch (resulting in this test's removal as well), and the `Tool` class presence/the tool itself feel like a distinct matter. If it can actually be dropped, I'd suggest adding a corresponding `TODO` around the corresponding `Tool`-derived driver class.

Reply:
1) I don't think there are any issues with backward compatibility as we are removing a check.  If these checks where to 'fail', it can be potentially assumed that the user would make modifications so these objects would not have kernels.  Allowing for multiple kernels in the final binary is something we already support for other modes, so expanding for FPGA should only raise issues if runtime does not handle the additional AOCX as expected.

2) Yes, I would like to get rid of llvm-no-spir-kernal usage as it isn't being used anymore.  The changes would be done separately, as they are not insignificant.  I will add the TODO as requested.

Reply:
Added TODO.

Reply:
I'm sorry for not realizing/elaborating on this initially.
Is it expected that, with this patch, the incremental FPGA linking flow becomes fully usable? In the event it is, I would strongly suggest that a simple E2E test akin to the one removed is written for the new functionality.

From my current understanding of the feature, the following layout would seem sufficient: 1. A couple modules with a couple dummy kernels each, compiled down to FPGA archive level in the per-kernel device code split mode; 2. One of the modules containing the main function that calls the kernels and including the header with the other module's declarations. Whichever module would get compiled in the second command could be linked against the first archive in the same command. Something like:
```bash
clang++ -fsycl -fintelfpga -fsycl-link=image main-module-with-2-kernels -I/second-module-header.h -fsycl-device-code-split=per_kernel -o image.a
clang++ -fsycl -fintelfpga image.a second-module-impl.cpp -I/second-module-header.h -fsycl-device-code-split=per_kernel -o final.out
```

Does some part of such a layout go against the actual intention of the feature? In case it does, could you please point me to these misconceptions? In case it doesn't, I'd express surprise and suggest the following:
1. If this testing scenario is technically correct and supportable, but a simpler example would cover everything as well, having a simpler one as an E2E test should do.
2. If this testing scenario is what's intended to achieve, but won't be fully supported by this patch alone, E2E test with a simpler case (e.g. w/o device code split enabled). The failures for the more complex case would help us identify the potential next steps.

Reply:
The main intention of the feature is to allow for multiple AOCX binaries to be available in the final executable.  For instance, if any `aocx` is generated or brought in from an FPGA archive, that `aocx` binary is integrated.  Your example should work fine, AFAIU.

Reply:
Do you believe this test can be added in scope of the current PR?

Reply:
The tests in the driver area cover the functionality of having an external `aocx` in an archive that is linked accordingly (https://github.com/intel/llvm/blob/93fe95958bb2c98a326e95b8a4d095f8a51c2af7/clang/test/Driver/sycl-offload-intelfpga.cpp#L170) which is similar to your suggestion above.  I believe separately a more succinct test can be added that would check for runtime expectations but we would need additional information (as already requested) to move forward.

Reply:
Right. Additional information being the extended use cases/runtime handling specifics?

Reply:
Yup, that's right.